pkg/election/election.go

```diff
@@ -323,10 +345,27 @@ func (e *Election) notifyLeader(ctx context.Context, leaderInfo *CampaignerInfo)
 	}
 }
 
-func (e *Election) watchLeader(ctx context.Context, session *concurrency.Session, key string) {
+func (e *Election) watchLeader(ctx context.Context, session *concurrency.Session, key string, elec *concurrency.Election) {
 	e.l.Debug("watch leader key", zap.String("key", key))
+
+	e.campaignMu.Lock()
+	e.resignCh = make(chan struct{})
+	e.campaignMu.Unlock()
+
+	defer func() {
+		e.campaignMu.Lock()
+		e.resignCh = nil
+		e.campaignMu.Unlock()
+	}()
+
 	wch := e.cli.Watch(ctx, key)
+
 	for {
+		if e.evictLeader.Get() {
+			elec.Resign(ctx)
```

Comment:
do we need to log the returned `err` here?

Reply:
yes, need to add log when resign failed, done